File: src/features/location/locations_entry.R
source("renv/activate.R")
source("src/features/utils/utils.R")
library("dplyr")
library("stringr")
library("tidyr")

location_data <-  read.csv(snakemake@input[["location_data"]], stringsAsFactors = FALSE)
day_segments_labels <-  read.csv(snakemake@input[["day_segments_labels"]], stringsAsFactors = FALSE)
provider <- snakemake@params["provider"][["provider"]]
provider_key <- snakemake@params["provider_key"]

location_features  <-  data.frame(local_segment = character(), stringsAsFactors = FALSE)

if(!"FEATURES" %in% names(provider))
        stop(paste0("Provider config[LOCATION][PROVIDERS][", provider_key,"] is missing a FEATURES attribute in config.yaml"))

if(provider[["COMPUTE"]] == TRUE){
  code_path <- paste0("src/features/location/", provider[["SRC_FOLDER"]], "/main.R")  
  source(code_path)
  features_function <- match.fun(paste0(provider[["SRC_FOLDER"]], "_location_features"))
  day_segments <- day_segments_labels %>% pull(label)
  for (day_segment in day_segments){
    print(paste(rapids_log_tag,"Processing", provider_key, day_segment))

    features <- features_function(location_data, day_segment, provider)

    # Check all features names contain the provider key so they are unique
    features_names <- colnames(features %>% select(-local_segment))
    if(any(!grepl(paste0(".*(",str_to_lower(provider_key),").*"), features_names)))
      stop(paste("The name of all location features of", provider_key," must contain its name in lower case but the following don't [", paste(features_names[!grepl(paste0(".*(",str_to_lower(provider_key),").*"), features_names)], collapse = ", "), "]"))

    location_features <- merge(location_features, features, all = TRUE)
  }
} else {
  for(feature in provider[["FEATURES"]])
    location_features[,feature] <- NA
}

location_features <- location_features %>% separate(col = local_segment, 
                                  into = c("local_segment_label", "local_start_date", "local_start_time", "local_end_date", "local_end_time"),
                                  sep = "#", 
                                  remove = FALSE)

write.csv(location_features, snakemake@output[[1]], row.names = FALSE)

File: src/features/location/locations_entry.py
import pandas as pd
from importlib import import_module, util
from pathlib import Path

# import filter_data_by_segment from src/features/utils/utils.py
spec = util.spec_from_file_location("util", str(Path(snakemake.scriptdir).parent / "utils" / "utils.py"))
mod = util.module_from_spec(spec)
spec.loader.exec_module(mod)
filter_data_by_segment = getattr(mod,  "filter_data_by_segment")
rapids_log_tag = getattr(mod,  "rapids_log_tag")

location_data = pd.read_csv(snakemake.input["location_data"][0])
day_segments_labels = pd.read_csv(snakemake.input["day_segments_labels"], header=0)
mypath = snakemake.params["mypath"]
provider = snakemake.params["provider"]
provider_key = snakemake.params["provider_key"]
location_features = pd.DataFrame(columns=["local_segment"])

if "FEATURES" not in provider:
        raise ValueError("Provider config[LOCATION][PROVIDERS][{}] is missing a FEATURES attribute in config.yaml".format(provider_key))

if provider["COMPUTE"] == True:
        code_path = provider["SRC_FOLDER"] + ".main"
        feature_module = import_module(code_path)
        feature_function = getattr(feature_module,  provider["SRC_FOLDER"] + "_location_features")
        
        for day_segment in day_segments_labels["label"]:
                print("{} Processing {} {}".format(rapids_log_tag, provider_key, day_segment))
                features = feature_function(location_data, day_segment, provider, filter_data_by_segment=filter_data_by_segment)
                location_features = location_features.merge(features, how="outer")
else:
        for feature in provider["FEATURES"]:
                location_features[feature] = None

segment_colums = pd.DataFrame()
segment_colums[["local_segment_label", "local_start_date", "local_start_time", "local_end_date", "local_end_time"]] = location_features["local_segment"].str.split(pat="#", expand=True)
for i in range(segment_colums.shape[1]):
        location_features.insert(1 + i, segment_colums.columns[i], segment_colums[segment_colums.columns[i]])
location_features.to_csv(snakemake.output[0], index=False)